test/hotspot/gtest/runtime/test_mutex_rank.cpp

```diff
@@ -0,0 +1,152 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/mutex.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "unittest.hpp"
+
+#ifdef ASSERT
+
+const int rankA = 50;
+
+TEST_OTHER_VM(MutexRank, mutex_lock_rank_in_order) {
+  JavaThread* THREAD = JavaThread::current();
+  ThreadInVMfromNative invm(THREAD);
+
+  Mutex* mutex_rankA = new Mutex(rankA, "mutex_rankA", false, Mutex::_safepoint_check_always);
+  Mutex* mutex_rankA_plus_one = new Mutex(rankA + 1, "mutex_rankA_plus_one", false, Mutex::_safepoint_check_always);
+
+  mutex_rankA_plus_one->lock();
+  mutex_rankA->lock();
+  mutex_rankA->unlock();
+  mutex_rankA_plus_one->unlock();
+}
+
+TEST_VM_ASSERT_MSG(MutexRank, mutex_lock_rank_out_of_orderA,
+                   "Attempting to acquire lock mutex_rankA_plus_one/51 out of order with lock mutex_rankA/50 -- possible deadlock") {
+  JavaThread* THREAD = JavaThread::current();
+  ThreadInVMfromNative invm(THREAD);
+
+  Mutex* mutex_rankA = new Mutex(rankA, "mutex_rankA", false, Mutex::_safepoint_check_always);
+  Mutex* mutex_rankA_plus_one = new Mutex(rankA + 1, "mutex_rankA_plus_one", false, Mutex::_safepoint_check_always);
+
+  mutex_rankA->lock();
+  mutex_rankA_plus_one->lock();
+  mutex_rankA_plus_one->unlock();
+  mutex_rankA->unlock();
+}
+
+TEST_VM_ASSERT_MSG(MutexRank, mutex_lock_rank_out_of_orderB,
+                   "Attempting to acquire lock mutex_rankB/50 out of order with lock mutex_rankA/50 -- possible deadlock") {
+  JavaThread* THREAD = JavaThread::current();
+  ThreadInVMfromNative invm(THREAD);
+
+  Mutex* mutex_rankA = new Mutex(rankA, "mutex_rankA", false, Mutex::_safepoint_check_always);
+  Mutex* mutex_rankB = new Mutex(rankA, "mutex_rankB", false, Mutex::_safepoint_check_always);
+
+  mutex_rankA->lock();
+  mutex_rankB->lock();
+  mutex_rankB->unlock();
+  mutex_rankA->unlock();
+}
+
+TEST_OTHER_VM(MutexRank, mutex_trylock_rank_out_of_order) {
+  JavaThread* THREAD = JavaThread::current();
+  ThreadInVMfromNative invm(THREAD);
+
+  Mutex* mutex_rankA = new Mutex(rankA, "mutex_rankA", false, Mutex::_safepoint_check_always);
+  Mutex* mutex_rankA_plus_one = new Mutex(rankA + 1, "mutex_rankA_plus_one", false, Mutex::_safepoint_check_always);
+  Mutex* mutex_rankA_plus_two = new Mutex(rankA + 2, "mutex_rankA_plus_two", false, Mutex::_safepoint_check_always);
+
+  mutex_rankA_plus_two->lock();
+  mutex_rankA->lock();
+  mutex_rankA_plus_one->try_lock_without_rank_check();
+  mutex_rankA_plus_one->unlock();
+  mutex_rankA->unlock();
+  mutex_rankA_plus_two->unlock();
```

Comment:
I think you should add a test that _skip_rank_check is reset, ie. lock mutex_rankA_plus_one with plain try_lock and make sure it tests the rank.

Reply:
Sounds good, I'll add that.